### ISSUE TITLE
VBMS Documents not Available in Caseflow

### ISSUE THREAD

**Issue by araposo-tistatech** [department-of-veterans-affairs/caseflow#14081]
Attorneys and VLJs are indicating that Reader is not pulling all the documents from VBMS when staff is reviewing appeals. 

This appears to have begun around April 8th. 

Examples provided below: 
https://appeals.cf.ds.va.gov/queue/appeals/3314680 - the entirety of this veterans claims file was not showing in Reader. The attorney could not view any documents past October 2018, while VLJ White could see 30 additional documents. Attorney went to VBMS to view documents

https://appeals.cf.ds.va.gov/queue/appeals/3621614 - Case was reviewed by VLJ White and was missing documents. VLJ White went to VBMS to view all documents. 

https://appeals.cf.ds.va.gov/queue/appeals/3765285 - Reported on Tuesday, VLJ White did not have IHP available in Reader. IHP was available in VBMS. 

https://appeals.cf.ds.va.gov/queue/appeals/dd3c3db8-bb37-4ea1-8166-7863118ce5ff - VLJ LeBoff could not find IHP in Reader, however it was uploaded into VBMS in December 2019. 

## Acceptance criteria
- [ ] Ensure all VBMS documents are loading and displaying in Reader

### Background/context
[Slack Conversation](https://dsva.slack.com/archives/CJL810329/p1587486489492500)

**Comment by yoomlam** [comment]
TLDR: Users have tagged IHP docs in Reader, so problems may have gone away.

DocCount is shown on Case Details page.
Reader’s Total is shown in Reader.
VBMS and VVA counts are shown in eFolder.

Logged-in as myself:
Case | DocCount | Reader’s Total | VBMS docs | VVA docs
-- | -- | -- | -- | --
3314680 | 289 | 289 | 266 | 23
3621614 | 138 | 138 | 124 | 14
3765285 | 112 | 112 | 120 | 99
dd3c3db8 | 149 | 149 | 129 | 20

Logged-in as BVAJWHITE (doesn’t have access to eFolder!): can see same doc counts.

## [3314680](https://appeals.cf.ds.va.gov/queue/appeals/3314680)
I can see documents with receipt date after Oct 2018.

## [3621614](https://appeals.cf.ds.va.gov/queue/appeals/3621614)
Found IHP doc (with tags "IR Back" and "IR HTN")

## [3765285](https://appeals.cf.ds.va.gov/queue/appeals/3765285) 
Found IHP doc (with tags "SC-CD" and "DIC")

## [dd3c3db8](https://appeals.cf.ds.va.gov/queue/appeals/dd3c3db8-bb37-4ea1-8166-7863118ce5ff)
Has 2 IHP docs (4/8/2020 and 12/3/2019)

Note: No Vet ID shown in Reader!

Logged in as BVAESLEBOFF (doesn't have access to eFolder!), able to see all 149 docs in Reader.

**Comment by araposo-tistatech** [comment]
@yoomlam Can we explain the discrepancy in the doc count for 3765285 or are the document counts for VBMS & VVA Docs incorrect in the table?

**Comment by yoomlam** [comment]
@araposo-tistatech I double-checked those numbers and there is no discrepancy.  I must have gone to the wrong tab when typing in the numbers.

**Comment by yoomlam** [comment]
The doc count for Case 3621614 went up by 1 to 139 (125 + 14). At least the counts are consistent between Case Details page, Reader, and eFolder.
Document counts for the other cases have not changed.

**Comment by araposo-tistatech** [comment]
Per the Board, another user has encountered of the document count being different in Caseflow. 

In the situation the attorney M. Mahmoudi was working on this [appeal](https://appeals.cf.ds.va.gov/queue/appeals/00feb68a-c34a-479b-a654-44bc76282a00), while working the case Reader had the document count at around 311 documents. Upon completion of the decision drafting task and sending the decision to the judge (J. White) for review the judge saw and reviewed 425 documents in Reader.

When the case was routed back to the attorney for a re-write the attorney noticed there were over 100 documents that were not reviewed by him originally on the case. 

The Board believes there could be some sort of timing issue when loading the documents, but I do want to note that in the screenshots provided the document count in VBMS was 397 while Readers is 425. 

Can we continue looking into this? The Board is concerned this will cause a lack of trust in Caseflow.

**Comment by araposo-tistatech** [comment]
Could this be a cache issue?

**Comment by yoomlam** [comment]
> Could this be a cache issue?

I don't think so. From what I can tell in the code, only these 2 things are 'cached':
* EFolder document counts are cached: app/services/external_api/efolder_service.rb:8
* VBMS documents are "cached" (retrieved from VBMS then saved in S3) if it does not already exist in S3 ([code](https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/document.rb#L122)). This is triggered when VBMSService fetches the document ([code](https://github.com/department-of-veterans-affairs/caseflow/blob/b3441c2c90a4dd5a28eabea9024c4927efcba5ca/app/services/external_api/vbms_service.rb#L30)).

Additionally, for a document to be cached, it must first be retrieved. 426 docs were created in Caseflow's DB on these dates:
- [Mon, 23 Mar 2020, 421], 
- [Fri, 24 Apr 2020, 2], 
- [Sun, 26 Apr 2020, 3]

These docs are created by DocumentFetcher ([code](https://github.com/department-of-veterans-affairs/caseflow/blob/7e38376d2cda2c0d5c8ffe9a5097c54d89aa6d7c/app/services/document_fetcher.rb#L60)) once they have been retrieved. DocumentFetcher uses 2 document_services (EFolderService and VBMSService [code](https://github.com/department-of-veterans-affairs/caseflow/blob/7e38376d2cda2c0d5c8ffe9a5097c54d89aa6d7c/app/services/document_fetcher.rb#L77)) to do retrieval.

**Comment by yoomlam** [comment]
```ruby
appeal.treee :id,:status,:created_at,:assigned_at,:updated_at
                                         ┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
Appeal 16082 (evidence_submission) ───── │ ID     │ STATUS    │ CREATED_AT              │ ASSIGNED_AT             │ UPDATED_AT              │ CLOSED_AT               │
└── RootTask                             │ 314464 │ on_hold   │ 2019-08-07 15:06:14 UTC │ 2019-08-07 15:06:14 UTC │ 2019-08-07 15:06:14 UTC │                         │
    ├── DistributionTask                 │ 314465 │ completed │ 2019-08-07 15:06:14 UTC │ 2019-10-21 00:30:18 UTC │ 2020-04-14 18:49:46 UTC │ 2020-04-14 18:49:46 UTC │
    │   └── EvidenceSubmissionWindowTask │ 314466 │ completed │ 2019-08-07 15:06:14 UTC │ 2019-08-07 15:06:14 UTC │ 2019-10-21 00:30:18 UTC │ 2019-10-21 00:30:18 UTC │
    ├── JudgeAssignTask                  │ 826685 │ completed │ 2020-04-14 18:49:46 UTC │ 2020-04-14 18:49:46 UTC │ 2020-04-20 11:40:41 UTC │ 2020-04-20 11:40:41 UTC │
    └── JudgeDecisionReviewTask          │ 837873 │ assigned  │ 2020-04-20 11:40:41 UTC │ 2020-05-12 11:53:57 UTC │ 2020-05-12 11:53:57 UTC │                         │
        ├── AttorneyTask                 │ 837874 │ completed │ 2020-04-20 11:40:41 UTC │ 2020-04-20 11:40:41 UTC │ 2020-04-28 09:46:47 UTC │ 2020-04-28 09:46:47 UTC │
        └── AttorneyRewriteTask          │ 863547 │ completed │ 2020-04-29 14:30:06 UTC │ 2020-04-29 14:30:06 UTC │ 2020-05-12 11:53:57 UTC │ 2020-05-12 11:53:57 UTC │
                                         └────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```

Timeline:
* Mar 23: 421 documents were retrieved by Caseflow
* Apr 20 Judge assigns to atty
* Apr 21 Atty starts to view 309 docs on these dates:
 [Tue, 21 Apr 2020 00:00:00 UTC +00:00, 303],
 [Wed, 22 Apr 2020 00:00:00 UTC +00:00, 2],
 [Mon, 27 Apr 2020 00:00:00 UTC +00:00, 4]]
* 5 more documents were retrieved by Caseflow
[Fri, 24 Apr 2020, 2],
[Sun, 26 Apr 2020, 3]
* Apr 28 Atty completes AttorneyTask
* Apr 28 Judge views 32 docs
[[Wed, 29 Apr 2020 00:00:00 UTC +00:00, 22],
 [Tue, 28 Apr 2020 00:00:00 UTC +00:00, 10]]
* Apr 29 Atty assigned AttorneyRewriteTask
* May 6 Atty views 93 docs for the first time on these dates:
 [Wed, 06 May 2020 00:00:00 UTC +00:00, 21],
 [Mon, 11 May 2020 00:00:00 UTC +00:00, 71],
 [Tue, 12 May 2020 00:00:00 UTC +00:00, 1],
* May 12 Atty completed AttorneyRewriteTask

**Comment by araposo-tistatech** [comment]
Posted this to the wrong ticket: 

@hschallhorn now that we've opened new tickets, should we close this one out? Also, just a reminder of the following notes from Troy:

1. Logs show March 23 there were 421 documents downloaded but user claims on April 21 that not all documents were in Caseflow. Does Caseflow use what is in Caseflow storage at the time the user views the record in Caseflow to report the number of documents or does it make a call to VBMS at the point in time of user action? If the former, then an incorrect number in Caseflow would have nothing to do with VBMS.

2. I’ve been told that if a document isn’t viewed within 2 weeks of download, it is removed from Caseflow storage. The date of download (March 23) to the date the user tried to view (April 21) is more than two weeks. Is it possible another user viewed the 309 documents within two weeks of April 21, hence those persisted but the others weren’t viewed and removed from storage after 2 weeks (April 4)? And the user failed to hit the refresh button so the missing documents didn’t refresh into Caseflow storage, hence the lower # of docs?

**Comment by hschallhorn** [comment]
From Yoom:

> Something to consider when investigating: the `current_user` is passed to fetch documents:
> https://github.com/department-of-veterans-affairs/caseflow/blob/24ed3d364a718ed40ae765f7368f824d41d702af/app/services/document_fetcher.rb#L81

> VBMS doesn't use it:
> https://github.com/department-of-veterans-affairs/caseflow/blob/b3441c2c90a4dd5a28eabea9024c4927efcba5ca/app/services/external_api/vbms_service.rb#L33

> but eFolder does make use of the `user`:
> https://github.com/department-of-veterans-affairs/caseflow/blob/d3f2ab606e6fe9796d7b6c9c92642f87a6680001/app/services/external_api/efolder_service.rb#L35

> Questions to consider during investigation: 
> * What triggers the initial document fetch? Who is the `current_user`?
> * Does the `current_user` have access to all documents (when eFolder fetcher is used)?

**Comment by yoomlam** [comment]
Added [more notes in Part 3](https://hackmd.io/9DYl3EwdTqKCpALVbIWnQg#Part-3) -- nothing conclusive yet.
